packages/common/src/api/tan-query/jupiter/useSwapTokens.ts

```diff
@@ -0,0 +1,375 @@
+  updateJupiterAtaCreationFeePayer
+} from './utils'
+
+/**
+ * Hook for executing token swaps using Jupiter.
+ * Swaps any supported SPL token (or SOL) for another.
+ */
+export const useSwapTokens = () => {
+  const queryClient = useQueryClient()
+  const { solanaWalletService, reportToSentry, audiusSdk } =
+    useAudiusQueryContext()
+  const { data: user } = useGetCurrentUser({})
+
+  return useMutation<SwapTokensResult, Error, SwapTokensParams>({
+    mutationFn: async (params): Promise<SwapTokensResult> => {
+      const {
+        inputMint: inputMintUiAddress,
+        outputMint: outputMintUiAddress,
+        amountUi
+      } = params
+      const slippageBps = params.slippageBps ?? 50
+      const wrapUnwrapSol = params.wrapUnwrapSol ?? true
+
+      let quoteResult
+      let signature: string | undefined
+      const instructions: TransactionInstruction[] = []
+
+      try {
+        // ---------- 1. Initialize Dependencies & Wallet ----------
+        const [sdk, keypair] = await Promise.all([
+          audiusSdk(),
+          solanaWalletService.getKeypair()
+        ])
+
+        if (!keypair) {
+          return {
+            status: SwapStatus.ERROR,
+            error: {
+              type: SwapErrorType.WALLET_ERROR,
+              message: 'Wallet not initialised'
+            }
+          }
+        }
+        const userPublicKey = keypair.publicKey
+        const feePayer = await sdk.services.solanaRelay.getFeePayer()
+        const ethAddress = user?.wallet
+
+        // ---------- 2. Get Quote from Jupiter ----------
+        try {
+          quoteResult = await getJupiterQuoteByMint({
+            inputMint: inputMintUiAddress,
+            outputMint: outputMintUiAddress,
+            amountUi,
+            slippageBps,
+            swapMode: 'ExactIn',
+            onlyDirectRoutes: true
+          })
+        } catch (error: any) {
+          reportToSentry({
+            name: 'JupiterSwapQuoteError',
+            error,
+            feature: Feature.TanQuery,
+            additionalInfo: { params }
+          })
+          return {
+            status: SwapStatus.ERROR,
+            error: {
+              type: SwapErrorType.QUOTE_FAILED,
+              message: error?.message ?? 'Failed to get swap quote'
+            }
+          }
+        }
+
+        // ---------- 3. Prepare Transaction Instructions ----------
+        const inputTokenConfig =
+          USER_BANK_MANAGED_TOKENS[inputMintUiAddress.toUpperCase()]
+        const outputTokenConfig =
+          USER_BANK_MANAGED_TOKENS[outputMintUiAddress.toUpperCase()]
+
+        let sourceAtaForJupiter: PublicKey | undefined
+        let isSourceAtaTemporary = false
+
+        // --- 3a. Handle Input Token (if user bank managed) ---
+        const isInputUserBankManaged = !!(
+          inputTokenConfig &&
+          ethAddress &&
+          inputMintUiAddress.toUpperCase() !== 'SOL'
+        )
+        if (isInputUserBankManaged) {
```

Comment:
I'm not sure I follow this logic - when wouldn't it be user bank managed? And why is that based on mint?

Seems to me based on usage that you probably want a parameter that lets the caller say "send from my user bank" or something

Reply:
there's a bit of premature behavior here that may not be needed, i guess do we ever see ourselves moving to an ATA based system?

Reply:
I think even if we do, we would want to signal that's what we want rather than implying it via the mints chosen. and we can always reference this commit when we want to add it!